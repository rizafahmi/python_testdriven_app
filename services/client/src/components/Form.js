import React from 'react'
import { Redirect } from 'react-router-dom'

const Form = props => {
  if (props.isAuthenticated) {
    return <Redirect to='/' />
  }
  return (
    <div>
      <h1>{props.formType}</h1>
      <form onSubmit={event => props.handleUserFormSubmit(event)}>
        {props.formType === 'Register' && (
          <div className='form-group'>
            <input
              className='form-control input-lg'
              name='username'
              type='text'
              placeholder='Enter a username'
              required
              value={props.formData.username}
              onChange={props.handleFormChange}
            />
          </div>
        )}
        <div className='form-group'>
          <input
            className='form-control input-lg'
            name='email'
            type='email'
            placeholder='Enter an email address'
            required
            value={props.formData.email}
            onChange={props.handleFormChange}
          />
        </div>
        <div className='form-group'>
          <input
            className='form-control input-lg'
            name='password'
            type='password'
            placeholder='Enter a password'
            required
            value={props.formData.password}
            onChange={props.handleFormChange}
          />
        </div>
        <input
          type='submit'
          className='btn btn-warning btn-lg btn-block'
          value='Submit'
        />
      </form>
    </div>
  )
}

export default Form
