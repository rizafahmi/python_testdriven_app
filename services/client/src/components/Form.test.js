import React from 'react'
import { shallow } from 'enzyme'
import renderer from 'react-test-renderer'

import Form from './Form'

const formData = {
  username: '',
  email: '',
  password: ''
}

test('Register form renders properly', () => {
  const component = <Form formType={'Register'} formData={formData} />
  const wrapper = shallow(component)
  const h1 = wrapper.find('h1')
  expect(h1.length).toBe(1)
  expect(h1.get(0).props.children).toBe('Register')

  const formGroup = wrapper.find('.form-group')
  expect(formGroup.length).toBe(3)
  expect(formGroup.get(0).props.children.props.name).toBe('username')
  expect(formGroup.get(0).props.children.props.value).toBe('')
})

test('Login form renders properly', () => {
  const component = <Form formType={'Login'} formData={formData} />
  const wrapper = shallow(component)
  const h1 = wrapper.find('h1')
  expect(h1.length).toBe(1)
  expect(h1.get(0).props.children).toBe('Login')

  const formGroup = wrapper.find('.form-group')
  expect(formGroup.length).toBe(2)
  expect(formGroup.get(0).props.children.props.name).toBe('email')
  expect(formGroup.get(0).props.children.props.value).toBe('')
})

test('Registered user redirect properly', () => {
  const component = (
    <Form formType={'Register'} formData={formData} isAuthenticated />
  )
  const wrapper = shallow(component)
  expect(wrapper.find('Redirect')).toHaveLength(1)
})

test('Login user redirect properly', () => {
  const component = (
    <Form formType={'Login'} formData={formData} isAuthenticated />
  )
  const wrapper = shallow(component)
  expect(wrapper.find('Redirect')).toHaveLength(1)
})

test('Register form submit properly', () => {
  const testValue = {
    formType: 'Register',
    formData: {
      username: '',
      email: '',
      password: ''
    },
    handleUserFormSubmit: jest.fn(),
    handleFormChange: jest.fn(),
    isAuthenticated: false
  }
  const wrapper = shallow(<Form {...testValue} />)
  const input = wrapper.find('input[type="text"]')
  expect(testValue.handleFormChange).toHaveBeenCalledTimes(0)
  input.simulate('change')
  expect(testValue.handleFormChange).toHaveBeenCalledTimes(1)
  expect(testValue.handleUserFormSubmit).toHaveBeenCalledTimes(0)
  wrapper.find('form').simulate('submit', testValue.formData)
  expect(testValue.handleUserFormSubmit).toHaveBeenCalledWith(
    testValue.formData
  )
  expect(testValue.handleUserFormSubmit).toHaveBeenCalledTimes(1)
})

test('Login form submit properly', () => {
  const testValue = {
    formType: 'Login',
    formData: {
      email: '',
      password: ''
    },
    handleUserFormSubmit: jest.fn(),
    handleFormChange: jest.fn(),
    isAuthenticated: false
  }
  const wrapper = shallow(<Form {...testValue} />)
  const input = wrapper.find('input[type="email"]')
  expect(testValue.handleFormChange).toHaveBeenCalledTimes(0)
  input.simulate('change')
  expect(testValue.handleFormChange).toHaveBeenCalledTimes(1)
  expect(testValue.handleUserFormSubmit).toHaveBeenCalledTimes(0)
  wrapper.find('form').simulate('submit', testValue.formData)
  expect(testValue.handleUserFormSubmit).toHaveBeenCalledWith(
    testValue.formData
  )
  expect(testValue.handleUserFormSubmit).toHaveBeenCalledTimes(1)
})

test('Register snapshot', () => {
  const component = <Form formType={'Register'} formData={formData} />
  const tree = renderer.create(component).toJSON()
  expect(tree).toMatchSnapshot()
})

test('Login snapshot', () => {
  const component = <Form formType={'Login'} formData={formData} />
  const tree = renderer.create(component).toJSON()
  expect(tree).toMatchSnapshot()
})
