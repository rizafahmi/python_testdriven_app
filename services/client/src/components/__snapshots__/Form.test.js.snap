// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login snapshot 1`] = `
<div>
  <h1>
    Login
  </h1>
  <form
    onSubmit={[Function]}
  >
    <div
      className="form-group"
    >
      <input
        className="form-control input-lg"
        name="email"
        onChange={undefined}
        placeholder="Enter an email address"
        required={true}
        type="email"
        value=""
      />
    </div>
    <div
      className="form-group"
    >
      <input
        className="form-control input-lg"
        name="password"
        onChange={undefined}
        placeholder="Enter a password"
        required={true}
        type="password"
        value=""
      />
    </div>
    <input
      className="btn btn-warning btn-lg btn-block"
      type="submit"
      value="Submit"
    />
  </form>
</div>
`;

exports[`Register snapshot 1`] = `
<div>
  <h1>
    Register
  </h1>
  <form
    onSubmit={[Function]}
  >
    <div
      className="form-group"
    >
      <input
        className="form-control input-lg"
        name="username"
        onChange={undefined}
        placeholder="Enter a username"
        required={true}
        type="text"
        value=""
      />
    </div>
    <div
      className="form-group"
    >
      <input
        className="form-control input-lg"
        name="email"
        onChange={undefined}
        placeholder="Enter an email address"
        required={true}
        type="email"
        value=""
      />
    </div>
    <div
      className="form-group"
    >
      <input
        className="form-control input-lg"
        name="password"
        onChange={undefined}
        placeholder="Enter a password"
        required={true}
        type="password"
        value=""
      />
    </div>
    <input
      className="btn btn-warning btn-lg btn-block"
      type="submit"
      value="Submit"
    />
  </form>
</div>
`;
